pkg/services/ngalert/api/api_testing_test.go

```diff
@@ -0,0 +1,284 @@
+		rc := &models2.ReqContext{
+			Context: &web.Context{
+				Req: &http.Request{},
+			},
+			SignedInUser: &models2.SignedInUser{
+				OrgId: 1,
+			},
+		}
+
+		t.Run("should return 401 if user cannot query a data source", func(t *testing.T) {
+			data1 := models.GenerateAlertQuery()
+			data2 := models.GenerateAlertQuery()
+
+			ac := acMock.New().WithPermissions([]*accesscontrol.Permission{
+				{Action: datasources.ActionQuery, Scope: datasources.ScopeProvider.GetResourceScopeUID(data1.DatasourceUID)},
+			})
+
+			srv := &TestingApiSrv{
+				accessControl: ac,
+			}
+
+			response := srv.RouteTestGrafanaRuleConfig(rc, definitions.TestRulePayload{
+				Expr: "",
+				GrafanaManagedCondition: &models.EvalAlertConditionCommand{
+					Condition: data1.RefID,
+					Data:      []models.AlertQuery{data1, data2},
+					Now:       time.Time{},
+				},
+			})
+
+			require.Equal(t, http.StatusUnauthorized, response.Status())
+		})
+
+		t.Run("should return 200 if user can query all data sources", func(t *testing.T) {
+			data1 := models.GenerateAlertQuery()
+			data2 := models.GenerateAlertQuery()
+
+			ac := acMock.New().WithPermissions([]*accesscontrol.Permission{
+				{Action: datasources.ActionQuery, Scope: datasources.ScopeProvider.GetResourceScopeUID(data1.DatasourceUID)},
+				{Action: datasources.ActionQuery, Scope: datasources.ScopeProvider.GetResourceScopeUID(data2.DatasourceUID)},
+			})
+
+			ds := &datasources.FakeCacheService{DataSources: []*models2.DataSource{
+				{Uid: data1.DatasourceUID},
+				{Uid: data2.DatasourceUID},
+			}}
+
+			evaluator := &eval.FakeEvaluator{}
+			var result []eval.Result
+			evaluator.EXPECT().ConditionEval(mock.Anything, mock.Anything, mock.Anything).Return(result, nil)
+
+			srv := &TestingApiSrv{
+				DatasourceCache: ds,
+				accessControl:   ac,
+				evaluator:       evaluator,
+			}
+
+			response := srv.RouteTestGrafanaRuleConfig(rc, definitions.TestRulePayload{
+				Expr: "",
+				GrafanaManagedCondition: &models.EvalAlertConditionCommand{
+					Condition: data1.RefID,
+					Data:      []models.AlertQuery{data1, data2},
+					Now:       time.Time{},
+				},
+			})
+
+			require.Equal(t, http.StatusOK, response.Status())
+
+			evaluator.AssertCalled(t, "ConditionEval", mock.Anything, mock.Anything, mock.Anything)
+		})
+	})
+
+	t.Run("when fine-grained access is disabled", func(t *testing.T) {
+		t.Run("should require user to be signed in", func(t *testing.T) {
+			data1 := models.GenerateAlertQuery()
+			ac := acMock.New().WithDisabled()
+
+			ds := &datasources.FakeCacheService{DataSources: []*models2.DataSource{
+				{Uid: data1.DatasourceUID},
+			}}
+
+			evaluator := &eval.FakeEvaluator{}
+			var result []eval.Result
+			evaluator.EXPECT().ConditionEval(mock.Anything, mock.Anything, mock.Anything).Return(result, nil)
+
+			srv := &TestingApiSrv{
+				DatasourceCache: ds,
+				accessControl:   ac,
+				evaluator:       evaluator,
+			}
```

Comment:
Each test here repeats some variant or subset of this same setup code. Is it worth genericizing this and extracting it to a SUT builder method that passes the configured mock in, like `func createTestingApiSrvSut(ac api.AccessControl) *TestingApiSrv`? This could help test readability by eliminating a lot of boilerplate out of each test.

Reply:
One alternative pattern I've seen (not that we strictly need to use it here) is to add getters to the SUT which return the dependency interfaces:

```
func (t *TestingApiSrv) evaluator() eval.Evaluator {
    return t.evaluator
}
```

then, the tests can use a shared builder and mock setup:

```
func createSut() *TestingApiSrv {
    ac := acMock.New().WithWhatever(...)
    eval := &eval.fakeEvaluator{}
    return &TestingApiSrv{...}
}
```

and when a test wants to access the mock directly for expectations, it gets the mock off the SUT and casts down, this is safe as the test knows which deps are mocked anyway:

```
sut := createSut()

sut.MethodToBeTested()

sut.accessControl().(*eval.fakeEvaluator).EXPECT()....
```

(some mock libraries will even offer a mock repository which lets you look up the concrete mock instance given an interface reference to skip yet another step. not sure if mockery supports that, but it's nice if so)

This specific pattern may be a little overkill for what we're trying to do here, but now that we're introducing mocks in a couple places, it's good to start thinking about patterns using them.

Reply:
Unfortunately, accesscontrol mock does not work that way. `ac := acMock.New().WithWhatever(...)` actually creates two structure (one by New() and another by WithWatever()). So injecting a mock in SUT and modifying it later would not work.